pkg/tsdb/parca/query.go

```diff
@@ -123,54 +128,161 @@ type CustomMeta struct {
+
+		for _, sample := range series.Samples {
+			timeField.Append(sample.Timestamp.AsTime())
+			valueField.Append(sample.Value)
+		}
+
+		fields = append(fields, valueField)
+		frame.Fields = fields
+	}
+
+	return frames
+}
+
+func arrowToNestedSetDataFrame(flamegraph *v1alpha1.FlamegraphArrow) *data.Frame {
 	frame := data.NewFrame("response")
+	frame.Meta = &data.FrameMeta{PreferredVisualization: "flamegraph"}
 
 	levelField := data.NewField("level", nil, []int64{})
 	valueField := data.NewField("value", nil, []int64{})
-	valueField.Config = &data.FieldConfig{Unit: normalizeUnit(tree.Unit)}
+	valueField.Config = &data.FieldConfig{Unit: normalizeUnit(flamegraph.Unit)}
 	selfField := data.NewField("self", nil, []int64{})
-	selfField.Config = &data.FieldConfig{Unit: normalizeUnit(tree.Unit)}
+	selfField.Config = &data.FieldConfig{Unit: normalizeUnit(flamegraph.Unit)}
 	labelField := data.NewField("label", nil, []string{})
 	frame.Fields = data.Fields{levelField, valueField, selfField, labelField}
 
-	walkTree(tree.Root, func(level int64, value int64, name string, self int64) {
+	arrowReader, err := ipc.NewReader(bytes.NewBuffer(flamegraph.GetRecord()))
+	if err != nil {
+		// TODO: Handle properly?
+		return nil
+	}
+	defer arrowReader.Release()
+
+	arrowReader.Next()
+	rec := arrowReader.Record()
+
+	arrowTable := array.NewTableFromRecords(arrowReader.Schema(), []arrow.Record{rec})
+	defer arrowTable.Release()
+
+	reader := array.NewTableReader(arrowTable, 0)
+	defer reader.Release()
+
+	fi, err := newFlamegraphIterator(reader)
+	if err != nil {
+		// TODO: Handle properly?
+		return nil
+	}
+
+	fi.iterate(func(name string, level, value, self int64) {
+		labelField.Append(name)
 		levelField.Append(level)
 		valueField.Append(value)
-		labelField.Append(name)
 		selfField.Append(self)
 	})
+
 	return frame
 }
 
-type Node struct {
-	Node  *v1alpha1.FlamegraphNode
-	Level int64
+const (
+	FlamegraphFieldMappingFile     = "mapping_file"
+	FlamegraphFieldLocationAddress = "location_address"
+	FlamegraphFieldFunctionName    = "function_name"
+	FlamegraphFieldChildren        = "children"
+	FlamegraphFieldCumulative      = "cumulative"
+	FlamegraphFieldFlat            = "flat"
+)
+
+type flamegraphIterator struct {
+	columnChildren         *array.List
+	columnChildrenValues   *array.Uint32
+	columnCumulative       func(i int) int64
+	columnMappingFile      *array.Dictionary
+	columnMappingFileDict  *array.Binary
+	columnFunctionName     *array.Dictionary
+	columnFunctionNameDict *array.Binary
+	columnLocationAddress  *array.Uint64
+
+	nameBuilder    *bytes.Buffer
+	addressBuilder *bytes.Buffer
+}
+
+func newFlamegraphIterator(tr *array.TableReader) (*flamegraphIterator, error) {
+	if !tr.Next() {
+		return nil, fmt.Errorf("table reader has no record")
+	}
+	rec := tr.Record()
+	schema := rec.Schema()
+
+	columnChildren := rec.Column(schema.FieldIndices(FlamegraphFieldChildren)[0]).(*array.List)
```

Comment:
Is there some doc about the arrow columns structure, how the flame graph is encoded in it? Didn't work with arrow before and even though I think I know how arrow works I am confused about what the columns mean in context of the flamegraph data.

Reply:
Sadly not. 
It's super close to the frames you use in your backend.
Each column is an array that encodes all values. 
To read a row you need to find the value in all arrays at the given index.

The children column is a list of lists. You can think of it as this: 
```json
[[0, 1, 2, 3], [4, 5], [6, 7], null, ...]
````
It's always indices pointing to other rows.
The first row is always our root row in the flame graph. Then this root row points to `[0, 1,2, 3]` in the example above. Once we've read the first row we know we need to walk these other rows to find the children of a row. 

I hope that explains the children column a bit better.

Reply:
Ah ok yeah now it makes more sense, thanks.